core/services/relay/evm/read/multieventtype.go

```diff
@@ -0,0 +1,215 @@
+			err = fmt.Errorf("no event queries provided: %w", err)
+		}
+	}
+
+	return seqIter, err
+}
+
+func multiEventTypeQueryWithoutErrorWrapping(ctx context.Context, lp logpoller.LogPoller, eventQueries []EventQuery, limitAndSort query.LimitAndSort) (iter.Seq2[string, commontypes.Sequence], error) {
+	if err := validateEventQueries(eventQueries); err != nil {
+		return nil, fmt.Errorf("error validating event queries: %w", err)
+	}
+
+	for _, eq := range eventQueries {
+		if err := eq.EventBinding.validateBound(eq.Address); err != nil {
+			return nil, err
+		}
+	}
+
+	allFilterExpressions := make([]query.Expression, 0, len(eventQueries))
+	for _, eq := range eventQueries {
+		var expressions []query.Expression
+
+		defaultExpressions := []query.Expression{
+			logpoller.NewAddressFilter(eq.Address),
+			logpoller.NewEventSigFilter(eq.EventBinding.hash),
+		}
+		expressions = append(expressions, defaultExpressions...)
+
+		remapped, remapErr := eq.EventBinding.remap(eq.Filter)
+		if remapErr != nil {
+			return nil, fmt.Errorf("error remapping filter: %w", remapErr)
+		}
+		expressions = append(expressions, remapped.Expressions...)
+
+		filterExpression := query.And(expressions...)
+
+		allFilterExpressions = append(allFilterExpressions, filterExpression)
+	}
+
+	eventQuery := query.Or(allFilterExpressions...)
+
+	queryName := createQueryName(eventQueries)
+
+	logs, err := lp.FilteredLogs(ctx, []query.Expression{eventQuery}, limitAndSort, queryName)
+	if err != nil {
+		return nil, wrapInternalErr(err)
+	}
+
+	seqIter, err := decodeMultiEventTypeLogsIntoSequences(ctx, logs, eventQueries)
+	if err != nil {
+		return nil, wrapInternalErr(err)
+	}
+
+	return seqIter, nil
+}
+
+func createQueryName(eventQueries []EventQuery) string {
+	queryName := ""
+	contractToEvents := map[string][]string{}
+	for _, eq := range eventQueries {
+		contractName := eq.EventBinding.contractName + "-" + eq.Address.String()
+
+		if _, exists := contractToEvents[contractName]; !exists {
+			contractToEvents[contractName] = []string{}
+		}
+		contractToEvents[contractName] = append(contractToEvents[contractName], eq.EventBinding.eventName)
+	}
+
+	contractNames := make([]string, 0, len(contractToEvents))
+	for contractName := range contractToEvents {
+		contractNames = append(contractNames, contractName)
+	}
+
+	sort.Strings(contractNames)
+
+	for _, contractName := range contractNames {
+		queryName += contractName + "-"
+		for _, event := range contractToEvents[contractName] {
+			queryName += event + "-"
+		}
+	}
+
+	queryName = strings.TrimSuffix(queryName, "-")
+	return queryName
+}
+
+func validateEventQueries(eventQueries []EventQuery) error {
+	duplicateCheck := map[common.Hash]EventQuery{}
+	for _, eq := range eventQueries {
+		if eq.EventBinding == nil {
+			return errors.New("event binding is nil")
+		}
+
+		if eq.SequenceDataType == nil {
+			return errors.New("sequence data type is nil")
+		}
+
+		if _, exists := duplicateCheck[eq.EventBinding.hash]; exists {
```

Comment:
Can you add a TODO here, don't want to bloc you further, but this shouldn't really be an error but a supported behaviour

Reply:
sure - done